File: frontend/src/constants/referenceIds.js
// Generate reference IDs from REF-1000 to REF-9999 (9,000 total IDs)
export const DEMO_REFERENCE_IDS = Array.from({ length: 9000 }, (_, i) => {
  const num = 1000 + i;
  return `REF-${num.toString().padStart(4, '0')}`;
});

export const normalizeDemoReferenceId = (value) => {
  if (value == null) return null;

  const cleaned = String(value).trim().toUpperCase();
  if (!cleaned) return null;

  const match = cleaned.match(/^(?:CRM|REF)[-_\s]?(\d{4})$/);
  if (!match) return null;

  const normalized = `REF-${match[1]}`;

  // Check if the number is in valid range (1000-9999)
  const num = parseInt(match[1], 10);
  return (num >= 1000 && num <= 9999) ? normalized : null;
};

File: backend/app/constants/reference_ids.py
"""Demo reference IDs used by the chatbot intake flow."""

import re
from typing import Optional


# Generate reference IDs from REF-1000 to REF-9999 (8,999 total IDs)
DEMO_REFERENCE_IDS = [f"REF-{i:04d}" for i in range(1000, 10000)]

_DEMO_REFERENCE_SET = set(DEMO_REFERENCE_IDS)


def normalize_demo_reference_id(value: Optional[str]) -> Optional[str]:
    """Normalize demo reference IDs to the canonical REF-#### format."""
    if value is None:
        return None

    cleaned = str(value).strip().upper()
    if not cleaned:
        return None

    match = re.fullmatch(r"(?:CRM|REF)[-_\s]?(\d{4})", cleaned)
    if not match:
        return None

    normalized = f"REF-{match.group(1)}"
    if normalized not in _DEMO_REFERENCE_SET:
        return None

    return normalized
